### PR TITLE
Use LLVMBuider to simplify generating LLVM code - Part 2

### PR DESCRIPTION
**Pull request by tungld**
Thi patch continues PR #1498 and creates a helper function IfThenElse in LLVMBuilder.

This patch also adds a check to verify non-negative dimensions of the input tensors, reported here #1506

Signed-off-by: Tung D. Le <tung@jp.ibm.com>

### CODE REVIEW

File: src/Dialect/Mlir/DialectBuilder.hpp [hunk 1]
```diff
@@ -336,6 +345,21 @@ struct LLVMBuilder final : DialectBuilder {
...
+  /// llvm.condBr cond, ^thenBlock, ^elseBlock
+  /// ^thenBlock:
+  ///   thenBody
+  /// ^elseBlock:
+  ///   elseBody
+  /// ^mainBlock
+  ///   ...
+  /// ```
+  void ifThenElse(mlir::function_ref<mlir::Value(LLVMBuilder &createLLVM)> cond,
+      mlir::function_ref<void(LLVMBuilder &createLLVM)> thenFn,
```
Comment (AlexandreEichenberger) [inline src/Dialect/Mlir/DialectBuilder.hpp:361]: Maybe you could create a special name for this type? for example `using` like here `  using ApiRegistry = std::map<RuntimeAPI::API, RuntimeAPI>;`

File: src/Conversion/KrnlToLLVM/ConvertKrnlToLLVM.cpp [hunk 2]
```diff
@@ -267,36 +267,21 @@ void genSignatureFunction(ModuleOp &module,
...
-    create.llvm.store(noep, numOfEntryPointsPtr);
-    b.create<LLVM::BrOp>(loc, ValueRange(), endBlock);
-
-    // Emit code for the false block: do nothing.
-    b.setInsertionPointToStart(falseBlock);
-    b.create<LLVM::BrOp>(loc, ValueRange(), endBlock);
-
-    // Emit code for the end block to return the entry point array.
-    b.setInsertionPointToStart(endBlock);
+    create.llvm.ifThenElse(/*cond=*/
```
Comment (AlexandreEichenberger) [inline src/Conversion/KrnlToLLVM/ConvertKrnlToLLVM.cpp:270]: great saving in number of lines and so much more readable

File: src/Dialect/Mlir/DialectBuilder.cpp [hunk 3]
```diff
@@ -942,4 +953,47 @@ FlatSymbolRefAttr LLVMBuilder::getOrInsertSymbolRef(ModuleOp module,
...
+  // Emit code for the IF block.
+  b.setInsertionPointToEnd(ifBlock);
+  Value condVal = cond(createLLVM);
+
+  // Branch the block into the THEN and ELSE blocks.
+  condBr(condVal, thenBlock, {}, elseBlock, {});
+
+  // Emit code for the THEN block.
+  b.setInsertionPointToStart(thenBlock);
+  thenFn(createLLVM);
```
Comment (AlexandreEichenberger) [inline src/Dialect/Mlir/DialectBuilder.cpp:983]: You should create a new builder with the local builder, see what is done in SCF
```C++
          SCFBuilder scfBuilder(childBuilder, childLoc);
```
It was the recommendation of MLIR to build a local builder that is specific to the loc.

File: src/Dialect/Mlir/DialectBuilder.cpp [hunk 4]
```diff
@@ -942,4 +953,47 @@ FlatSymbolRefAttr LLVMBuilder::getOrInsertSymbolRef(ModuleOp module,
   return SymbolRefAttr::get(b.getContext(), funcName);
 }
 
+void LLVMBuilder::ifThenElse(
+    mlir::function_ref<Value(LLVMBuilder &createLLVM)> cond,
+    mlir::function_ref<void(LLVMBuilder &createLLVM)> thenFn,
+    mlir::function_ref<void(LLVMBuilder &createLLVM)> elseFn) const {
```
Comment (AlexandreEichenberger) [inline src/Dialect/Mlir/DialectBuilder.cpp:959]: Can you create a custom type using a "using"

File: src/Dialect/Mlir/DialectBuilder.cpp (hunk 3 above)
Comment (tungld) [reply src/Dialect/Mlir/DialectBuilder.cpp:983]: SCF is a bit different here. We are using the builder of `scf.IfOp` that uses `function_ref`, e.g. `function_ref<void(OpBuilder &, Location)> thenBuilder`, so we need to use the local location. But, if we use another builder of `scf.IfOp` that does not use `function_ref` but `setInsertionPointToStart `, the local location is not necessary, e.g. https://github.com/llvm/llvm-project/blob/main/mlir/lib/Dialect/SparseTensor/Transforms/Sparsification.cpp#L746.

IfThenElse here for LLVM is similar to the later one, where we just emit a sequence of ops to represent ifThenElse via LLVM::CondBrOp. LLVM dialect does not have ifOp.

File: src/Dialect/Mlir/DialectBuilder.hpp (hunk 1 above)
Comment (tungld) [reply src/Dialect/Mlir/DialectBuilder.hpp]: Yes, I created these aliases inside LLVMBuilder:
```C
using voidFuncRef = mlir::function_ref<void(LLVMBuilder &createLLVM)>;
using valueFuncRef = mlir::function_ref<mlir::Value(LLVMBuilder &createLLVM)>;
```
Please let me know if you prefer another name. Thank!

File: src/Dialect/Mlir/DialectBuilder.cpp (hunk 4 above)
Comment (tungld) [reply src/Dialect/Mlir/DialectBuilder.cpp]: Done. Thanks!

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, tx for all the changes

**Pull request merged**